Add API to include backgrounds when printing

rr-codes opened the pull request:
#### eee10982d72fa04055ce4c123a17d77e36e71946
<pre>
Add API to include backgrounds when printing
<a href="https://bugs.webkit.org/show_bug.cgi?id=244270">https://bugs.webkit.org/show_bug.cgi?id=244270</a>
rdar://90055962

Reviewed by Tim Horton and Aditya Keerthi.

Adds an API to allow for customization of if backgrounds should be included
when printing or not.

Also changes the default value of this option to `NO` on iOS, for consistency
with macOS. We are using a linked-on-or-after check to avoid breakage.

* Source/WebKit/UIProcess/API/Cocoa/WKPreferences.h:
* Source/WebKit/UIProcess/API/Cocoa/WKPreferences.mm:
(-[WKPreferences encodeWithCoder:]):
(-[WKPreferences initWithCoder:]):
(-[WKPreferences setShouldPrintBackgrounds:]):
(-[WKPreferences shouldPrintBackgrounds]):
(-[WKPreferences _setShouldPrintBackgrounds:]):
(-[WKPreferences _shouldPrintBackgrounds]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _setupPageConfiguration:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfiguration.mm:
(-[WKWebViewConfiguration init]):
(-[WKWebViewConfiguration copyWithZone:]):
(-[WKWebViewConfiguration _printsBackgrounds]):
(-[WKWebViewConfiguration _setPrintsBackgrounds:]):
* Source/WebKitLegacy/mac/WebView/WebPreferenceKeysPrivate.h:
* Source/WebKitLegacy/mac/WebView/WebPreferences.h:
* Source/WebKitLegacy/mac/WebView/WebPreferences.mm:
(-[WebPreferences setShouldPrintBackgrounds:]):
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKit/WKPreferences.cpp:
(TestWebKitAPI::TEST):
* Tools/TestWebKitAPI/Tests/WebKit/backgroundcolor.html: Added.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/Copying.mm:
(TEST):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebViewPrintFormatter.mm:
(TEST):

Canonical link: <a href="https://commits.webkit.org/253952@main">https://commits.webkit.org/253952@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8b53a87e1044910ab762c04152cee917890d3526

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/87521 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/31611 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/18315 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/96748 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/150475 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/91494 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/29974 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/26143 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/79640 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/91509 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/93140 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/68/builds/24221 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/74302 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/24048 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/79183 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/79394 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/67066 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/79358 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/27682 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/13227 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/73040 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/27635 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/14243 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/26002 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/2778 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/29325 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/37116 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/75868 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/29250 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/33521 "Passed tests") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/16819 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WTF/wtf/cocoa/RuntimeApplicationChecksCocoa.h:
```diff
@@ -40,6 +40,7 @@ enum class SDKAlignedBehavior {
     ConvertsInvalidURLsToBlank,
     DataURLFragmentRemoval,
     DecidesPolicyBeforeLoadingQuickLookPreview,
+    DefaultsToNotIncludingBackgroundsWhenPrinting,
```
`DefaultsToExcludingBackgroundsWhenPrinting`?

pxlcoder commented on Source/WebKit/UIProcess/API/Cocoa/WKPreferences.mm:
```diff
@@ -1019,12 +1033,12 @@ - (BOOL)_pageCacheSupportsPlugins
 - (BOOL)_shouldPrintBackgrounds
 {
-    return _preferences->shouldPrintBackgrounds();
+    return [self shouldPrintBackgrounds];
```
Nit: This can use dot notation.

rr-codes replied:
should `setShouldPrintBackgrounds` also use dot notation?

pxlcoder replied:
That would be good too. My understanding is that we prefer dot notation when possible (the notable exception being when using `RetainPtr`).

rr-codes replied:
also would It be `    return self.shouldPrintBackgrounds;` or without the `self`?

pxlcoder replied:
`return self.shouldPrintBackgrounds`

pxlcoder commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -472,7 +472,8 @@ - (void)_setupPageConfiguration:(Ref<API::PageConfiguration>&)pageConfiguration
     pageConfiguration->preferences()->setShouldRespectImageOrientation(!![_configuration _respectsImageOrientation]);
 #if !PLATFORM(MAC)
     // FIXME: Expose WKPreferences._shouldPrintBackgrounds on iOS, adopt it in all iOS clients, and remove this and WKWebViewConfiguration._printsBackgrounds.
```
This comment should be updated to reflect the fact that we have API now.

Can leave a FIXME with a new radar to track the SPI removal on iOS.

pxlcoder reviewed (commented):
We should also add a new API test case in `TEST(Coding, WKPreferences)`.

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/Copying.mm:
```diff
@@ -34,8 +34,10 @@
+    [a setShouldPrintBackgrounds:NO];
     [a setJavaScriptCanOpenWindowsAutomatically:YES];
 #else
+    [a setShouldPrintBackgrounds:YES];
```
Why are you setting different values on iOS / macOS?

(Same question further down)

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebViewPrintFormatter.mm:
```diff
@@ -102,33 +105,54 @@ - (void)_cleanPrintState;
 {
-    auto webView = adoptNS([[TestWKWebView alloc] initWithFrame:NSMakeRect(0, 0, 800, 600)]);
+    auto config = adoptNS([[WKWebViewConfiguration alloc] init]);
+    config.get().preferences.shouldPrintBackgrounds = NO;
```
Nit: `[config preferences].shouldPrintBackgrounds = NO`.

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebViewPrintFormatter.mm:
```diff
@@ -102,33 +105,54 @@ - (void)_cleanPrintState;
+    
+    runTest(NO);
+    
+    webView.get().configuration.preferences.shouldPrintBackgrounds = YES;
```
Nit: `[webView configuration].preferences.shouldPrintBackgrounds = YES`.

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKit/backgroundcolor.html:
```diff
@@ -0,0 +1,11 @@
+<body>
+<div>
+    hello
+  <div style="width: 100px; height: 100px; background-color: blue">
```
Do we need a new test page, or can you use the existing `red.html`?

rr-codes replied:
well that's a very aptly named and useful existing file lol, I shall use that

hortont424 commented on Source/WebKit/Shared/WebPreferencesDefaultValues.cpp:
```diff
@@ -54,6 +54,12 @@ bool defaultCSSOMViewScrollingAPIEnabled()
 
+bool defaultShouldPrintBackgrounds()
+{
+    static bool result = !linkedOnOrAfterSDKWithBehavior(SDKAlignedBehavior::DefaultsToExcludingBackgroundsWhenPrinting);
```
Someday we should convince Sam to make this a thing you can do straight from the YAML (since I see a bunch of them).

hortont424 commented on Source/WebKit/UIProcess/API/Cocoa/WKPreferences.h:
```diff
@@ -51,6 +51,11 @@ WK_CLASS_AVAILABLE(macos(10.10), ios(8.0))
+/*! @abstract A Boolean value indicating whether the web view should include backgrounds when printing.
+ @discussion The default value is `NO`.
+ */
+@property (nonatomic) BOOL shouldPrintBackgrounds;
```
Where's the availability macros?

hortont424 commented on Source/WebKitLegacy/mac/WebView/WebPreferences.h:
```diff
@@ -209,12 +209,10 @@ WEBKIT_CLASS_DEPRECATED_MAC(10_3, 10_14)
 */
 @property (nonatomic) BOOL autosaves;
 
-#if !TARGET_OS_IPHONE
```
This file is public API, I doubt you mean to be exposing new WebKitLegacy public API on iOS?

hortont424 replied:
I guess it's *not* public API on iOS...

hortont424 replied:
But it is on mac. But this is preprocessed out there? OK, nevermind!

rr-codes committed:
Add API to include backgrounds when printing
https://bugs.webkit.org/show_bug.cgi?id=244270
rdar://90055962

Reviewed by Tim Horton and Aditya Keerthi.

Adds an API to allow for customization of if backgrounds should be included
when printing or not.

Also changes the default value of this option to `NO` on iOS, for consistency
with macOS. We are using a linked-on-or-after check to avoid breakage.

* Source/WebKit/UIProcess/API/Cocoa/WKPreferences.h:
* Source/WebKit/UIProcess/API/Cocoa/WKPreferences.mm:
(-[WKPreferences encodeWithCoder:]):
(-[WKPreferences initWithCoder:]):
(-[WKPreferences setShouldPrintBackgrounds:]):
(-[WKPreferences shouldPrintBackgrounds]):
(-[WKPreferences _setShouldPrintBackgrounds:]):
(-[WKPreferences _shouldPrintBackgrounds]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _setupPageConfiguration:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfiguration.mm:
(-[WKWebViewConfiguration init]):
(-[WKWebViewConfiguration copyWithZone:]):
(-[WKWebViewConfiguration _printsBackgrounds]):
(-[WKWebViewConfiguration _setPrintsBackgrounds:]):
* Source/WebKitLegacy/mac/WebView/WebPreferenceKeysPrivate.h:
* Source/WebKitLegacy/mac/WebView/WebPreferences.h:
* Source/WebKitLegacy/mac/WebView/WebPreferences.mm:
(-[WebPreferences setShouldPrintBackgrounds:]):
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKit/WKPreferences.cpp:
(TestWebKitAPI::TEST):
* Tools/TestWebKitAPI/Tests/WebKit/backgroundcolor.html: Added.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/Copying.mm:
(TEST):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebViewPrintFormatter.mm:
(TEST):

Canonical link: https://commits.webkit.org/253952@main
